tests/Aspire.Dashboard.Tests/Utils/CertificateUtilTests.cs

```diff
@@ -0,0 +1,132 @@
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Xunit;
+
+namespace Aspire.Dashboard;
+
+public sealed class CertificateUtilTests
+{
+    [Theory]
+    [InlineData("testCert.pfx", X509ContentType.Pkcs12)]
+    [InlineData("https-dsa.pem", X509ContentType.Cert)]
+    public void TestCertificateTypes(string fileName, X509ContentType contentType)
+    {
+        var filePath = TestCertificateLoader.GetCertPath(fileName);
+
+        Assert.Equal(contentType, X509Certificate2.GetCertContentType(filePath));
+    }
+
+    [Fact]
+    public void GetFileCertificate_Pkcs12_CorrectPassword()
+    {
+        var certificates = CertificateUtil.GetFileCertificate(
+            filePath: TestCertificateLoader.GetCertPath("testCert.pfx"),
+            password: "testPassword",
+            logger: NullLogger.Instance);
+
+        Assert.NotNull(certificates);
+
+        var cert = Assert.Single(certificates.Cast<X509Certificate>());
+
+        Assert.NotNull(cert);
+        Assert.Equal("CN=localhost", cert.Subject);
+    }
+
+    [Fact]
+    public void GetFileCertificate_Pkcs12_IncorrectPassword()
+    {
+        Assert.Throws<CryptographicException>(() => CertificateUtil.GetFileCertificate(
+            filePath: TestCertificateLoader.GetCertPath("testCert.pfx"),
+            password: "wrongPassword",
+            logger: NullLogger.Instance));
+    }
+
+    [Fact]
+    public void GetFileCertificate_Cert_WithRedundantPassword()
+    {
+        var filePath = TestCertificateLoader.GetCertPath("https-dsa.pem");
+
+        using MockLogger logger = new()
+        {
+            $"Resource service certificate {filePath} has type {X509ContentType.Cert} which does not support passwords, yet a password was configured. The certificate password will be ignored."
+        };
+
+        var certificates = CertificateUtil.GetFileCertificate(
+            filePath: filePath,
+            password: "testPassword",
+            logger: logger);
+
+        Assert.NotNull(certificates);
+
+        var cert = Assert.Single(certificates.Cast<X509Certificate>());
+
+        Assert.NotNull(cert);
+        Assert.Equal("OU=Development, O=Contoso, L=Alexandria, S=Virginia, C=US", cert.Subject);
+    }
+
+    [Fact]
+    public void GetFileCertificate_Cert()
+    {
+        var filePath = TestCertificateLoader.GetCertPath("https-dsa.pem");
+
+        var certificates = CertificateUtil.GetFileCertificate(
+            filePath: filePath,
+            password: null, // No password required for this file
+            logger: null!); // Should not be called, and a sneaky null validates this
+
+        Assert.NotNull(certificates);
+
+        var cert = Assert.Single(certificates.Cast<X509Certificate>());
+
+        Assert.NotNull(cert);
+        Assert.Equal("OU=Development, O=Contoso, L=Alexandria, S=Virginia, C=US", cert.Subject);
+    }
+
+    private sealed class MockLogger : ILogger, IDisposable, IEnumerable
```

Comment:
You can use TestSink for this: https://github.com/dotnet/aspire/blob/9b5051080379d62b2066ab334be0e35064f4fdf9/tests/Aspire.Dashboard.Tests/Model/ApplicationsSelectHelpersTests.cs#L94-L103

Reply:
That code would be simpler if it used this `MockLogger` class.

`MockLogger` also throws as soon as an unexpected write occurs, so you get the stack involved.

Reply:
That's still a lot of extra code to complicate test methods when compared with:

```c#
using var logger = new MockLogger() { "Expected line" };
...
```